### PR TITLE
Use filesystem to lookup tsconfig files

### LINKED ISSUE #1312: Use filesystem to lookup tsconfig files

### PR DESCRIPTION
**Pull request by saberduck**
Fixes #1312

### CODE REVIEW

File: its/plugin/tests/src/test/java/com/sonar/javascript/it/plugin/Tests.java [hunk 1]
```diff
@@ -44,6 +44,7 @@
   EslintReportTest.class,
   MetricsTest.class,
   MinifiedFilesTest.class,
+  MultiTsconfigTest.class,
```
Comment (vilchik-elena) [inline its/plugin/tests/src/test/java/com/sonar/javascript/it/plugin/Tests.java]: 👍

File: its/plugin/tests/src/test/resources/ts-eslint-based-rules.xml [hunk 2]
```diff
@@ -8,5 +8,15 @@
...
       <priority>INFO</priority>
     </rule>
+    <rule>
+      <repositoryKey>typescript</repositoryKey>
+      <key>S1764</key>
+      <priority>INFO</priority>
+    </rule>
+    <rule>
+      <repositoryKey>typescript</repositoryKey>
+      <key>S3801</key>
```
Comment (vilchik-elena) [inline its/plugin/tests/src/test/resources/ts-eslint-based-rules.xml]: why this rule is added?

File: its/plugin/tests/src/test/resources/ts-eslint-based-rules.xml [hunk 3]
```diff
@@ -8,5 +8,15 @@
       <key>S3923</key> <!-- all duplicate branches -->
       <priority>INFO</priority>
     </rule>
+    <rule>
+      <repositoryKey>typescript</repositoryKey>
+      <key>S1764</key>
```
Comment (vilchik-elena) [inline its/plugin/tests/src/test/resources/ts-eslint-based-rules.xml]: this rule is not yet defined for TS, how does it work??

File: sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/eslint/TypeScriptSensor.java [hunk 4]
```diff
@@ -91,11 +95,16 @@ protected void analyze(InputFile file, SensorContext context) {
...
     if (tsconfigs == null) {
       FileSystem fs = context.fileSystem();
-      tsconfigs = new ArrayList<>();
-      fs.inputFiles(fs.predicates().hasRelativePath("tsconfig.json")).forEach(file -> tsconfigs.add(file.absolutePath()));
+      Path baseDir = fs.baseDir().toPath();
+      try (Stream<Path> files = Files.walk(baseDir)) {
+        return files
+          .filter(p -> p.endsWith("tsconfig.json"))
+          .map(p -> p.toAbsolutePath().toString())
+          .collect(Collectors.toList());
```
Comment (vilchik-elena) [inline sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/eslint/TypeScriptSensor.java]: let's add log with number of detected tsconfigs at info level

File: its/plugin/tests/src/test/java/com/sonar/javascript/it/plugin/MultiTsconfigTest.java [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+      throw new IllegalStateException("Failed to run npm install");
+    }
+  }
+
+  @Test
+  public void test() {
+    SearchRequest request = new SearchRequest();
+    request.setComponentKeys(Collections.singletonList(PROJECT));
+    List<Issues.Issue> issuesList = newWsClient().issues().search(request).getIssuesList();
+    assertThat(issuesList).extracting(Issues.Issue::getLine).containsExactlyInAnyOrder(3, 4);
```
Comment (vilchik-elena) [inline its/plugin/tests/src/test/java/com/sonar/javascript/it/plugin/MultiTsconfigTest.java]: if we add log, let's assert it as well

File: its/plugin/projects/multi-tsconfig-test-project/src/foo/tsconfig.json [hunk 6]
```diff
@@ -0,0 +1,8 @@
+{
```
Comment (vilchik-elena) [inline its/plugin/projects/multi-tsconfig-test-project/src/foo/tsconfig.json]: let's add one more tsconfig but with non standard name like `foo_tsconfig.json` to make sure it's ignored (or may be it's better to have this check in UT)

File: its/plugin/tests/src/test/resources/ts-eslint-based-rules.xml (hunk 2 above)
Comment (saberduck) [reply its/plugin/tests/src/test/resources/ts-eslint-based-rules.xml]: indeed this is a mistake, it was working with JS rule, but I should drop this

File: sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/eslint/TypeScriptSensor.java (hunk 4 above)
Comment (saberduck) [reply sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/eslint/TypeScriptSensor.java]: I added the log on DEBUG level with actual paths

File: its/plugin/tests/src/test/java/com/sonar/javascript/it/plugin/MultiTsconfigTest.java (hunk 5 above)
Comment (saberduck) [reply its/plugin/tests/src/test/java/com/sonar/javascript/it/plugin/MultiTsconfigTest.java]: hmm, I am not sure we need to assert such logs. They are not facing the user.

File: its/plugin/projects/multi-tsconfig-test-project/src/foo/tsconfig.json (hunk 6 above)
Comment (saberduck) [reply its/plugin/projects/multi-tsconfig-test-project/src/foo/tsconfig.json]: added in UT

File: its/ruling/src/test/java/org/sonar/javascript/it/RulingTest.java [hunk 7]
```diff
@@ -56,8 +63,33 @@
...
+  }
+
+  @Parameters(name = "{0}")
+  public static Object[][] projects() {
+    return new Object[][]{
+      {"js-project", "js", "../sources/src"},
+      tsProject("vscode"),
+      tsProject("angular")
+    };
+//      return new Object[][]{{"console"}};
```
Comment (yassin-kammoun-sonarsource) [inline its/ruling/src/test/java/org/sonar/javascript/it/RulingTest.java]: Is it a reminder for debugging purpose ?

**Comment by vilchik-elena** [comment]
lgtm

**Pull request merged**